src/backend/gl/src/window/egl.rs

```diff
@@ -195,20 +219,26 @@ impl hal::Instance<crate::Backend> for Instance {
             Rwh::Xcb(handle) => handle.window as _,
             #[cfg(target_os = "android")]
             Rwh::Android(handle) => handle.a_native_window,
+            #[cfg(not(target_os = "android"))]
+            Rwh::Wayland(handle) => {
+                let window = wayland::egl_window_create(handle.surface, 1024, 768);
```

Comment:
this looks quite fishy to me. We aren't supposed to create a window here. We are supposed to get a window and initialize the GL context on it. `Rwh::Wayland` has the WL surface, and that's actually what we need here, not WL window being created. See:

> To obtain an on-screen rendering surface from a Wayland window, call
    eglCreatePlatformWindowSurface with a <dpy> that belongs to Wayland and
    a <native_window> that points to a `struct wl_egl_surface`.

Reply:
I saw that documentation as well and I think it's outdated. This is not creating a window. The documentation here is pretty sparse but you need to create a wayland egl window (https://jan.newmarch.name/Wayland/EGL/)

> An EGL window needs to be created from a Wayland surface, by the Wayland call wl_egl_window_create. This is then turned into an EGL drawing surface by the EGL call eglCreateWindowSurface. Drawing into this surface is generally done by an API such as OpenGL, and the choice of this has been set in the egl_context. This is set for the drawing surfaces by eglMakeCurrent.

Reply:
Here is a more explanatory article I found (https://ppaalanen.blogspot.com/2012/03/what-does-egl-do-in-wayland-stack.html): 

> The client passes the wl_display object to eglGetDisplay() and receives an EGLDisplay to be used with EGL calls. Then comes the trick that is denoted by the double-arrowed blue line from Wayland client to Mesa EGL in the diagram. The client calls the wayland-egl API (implemented in Mesa) function wl_egl_window_create() to get the native window handle. Normally you would just use the "real" native window object wl_surface (or an X11 Window if you were using X). The native window handle is used to create the EGLSurface EGL handle. **Wayland has this extra step and the wayland-egl API because a wl_surface carries no information of its size. When the EGL library allocates buffers, it needs to know the size, and wayland-egl API is the only way to tell that.**

Reply:
This link says "This project hasn't been updated since mid-2017."

> An EGL window needs to be created from a Wayland surface, by the Wayland call wl_egl_window_create

My expectation is that Winit calls that, or does something equivalent to this. gfx-rs shouldn't be creating any windows.

Reply:
That other link is even older, from 2012. It does sound peculiar indeed, but between the link and the actual spec for the specification we are using, I'd rather trust the spec. Note that the spec expects `wl_egl_surface`, but that function `wl_egl_window_create` returns `wl_egl_window`, so it's a clear mismatch.

Reply:
I'll look into this then - I saw the spec and when I tried that I got a segfault. And despite what the spec says, Weston, Wayland reference compositor uses `wl_egl_window_create` as well. I'll see if there are any other projects on github that pass in the surface directly.

https://github.com/wayland-project/weston/blob/master/clients/simple-egl.c#L365

Reply:
I'm having trouble finding `wl_egl_surface` at all. Perhaps, the spec is indeed outdated, and this is now called `wl_egl_window`? :/

Reply:
I actually missed that - my brain read `wl_egl_surface` and just `wl_surface` - and likewise I've never seen `wl_egl_surface` in [wayland-project/wayland](https://github.com/wayland-project/wayland).

What I can say for sure is most other projects I've used as a reference use `egl_window_create` on a surface. I think the name is misleading however, like the ppaalanen's blogpost article states a `wl_egl_window` is just a surface with size information (you can see [here](https://github.com/wayland-project/wayland/blob/master/egl/wayland-egl.c#L55) that the function to create a  wl_egl_window just wraps a surface with some width/height metadata).

I think, conceptually this code is correct. The bigger problem seems to be that the raw_window_handle that I have, that was created by `winit` has a `wl_surface` and a `wl_display`. The `wl_display` I get from the `raw_window_handle` is created by `winit`, not by `Instance::create`, and when I try to use this surface in `create_platform_surface_window`, wayland doesn't like this (logs from weston 8.0.0):

```
[21:01:54.017] libwayland: invalid object (6), type (zwp_relative_pointer_manager_v1), message attach(?oii)
[21:01:54.025] libwayland: error in client communication (pid 10906)
```

AFAICT, my options are:

1. Reinitialize the glcontext with this new `display`/window handle. This repeats the create function, and also could require an API change (`&mut self`).
2. Pass in the window handle during `create`. This is what I'm doing, via the version parameter, now. Again this breaks the API
3. Somehow figure this out without needing interaction with winit.

Reply:
> [21:01:54.017] libwayland: invalid object (6), type (zwp_relative_pointer_manager_v1), message attach(?oii)

This sounds like the object is bad. Is there a way to check if it's a good `wl_display`? I wonder if it's just a bug on `winit` side.